gen/wrap_proj.jl

```diff
@@ -0,0 +1,154 @@
+#=
+Run this file to regenerate `proj_c.jl` and `proj_common.jl`.
+
+It expects a PROJ install in the deps folder, run `build Proj4` in Pkg mode
+if these are not in place.
+
+The wrapped PROJ version and provided PROJ version should be kept in sync.
+So when updating the PROJBuilder provided version, also rerun this wrapper.
+This way we ensure that the provided library has the same functions available
+as the wrapped one. Furthermore this makes sure constants in `proj_common.jl`
+like `PROJ_VERSION_PATCH`, which are just literals, are correct.
+
+Several custom transformations are applied that should make using this package more convenient.
+- docstrings are added, created from PROJ Doxygen XML output
+- functions that return a Cstring are wrapped in unsafe_string to return a String
+- functions that return a Ptr{Cstring} are wrapped in unsafe_loadstringlist to return a Vector{String}
+- context arguments become keyword arguments, defaulting to C_NULL meaning default context
+
+These transformations are based on the code developed for GDAL.jl, see
+https://github.com/JuliaGeo/GDAL.jl/blob/master/gen/README.md for more information
+on how to construct the PROJ Doxygen XML file needed here.
+=#
+
+using Clang  # needs a post 0.9.1 release with #231 and #232
+using MacroTools
+using EzXML
+
+const xmlpath = joinpath(@__DIR__, "doxygen.xml")
+
+# several functions for building docstrings
+include(joinpath(@__DIR__, "doc.jl"))
+
+
+"""
+Custom rewriter for Clang.jl's C wrapper
+
+Gets called with all expressions in a header file, or all expressiong in a common file.
+If available, it adds docstrings before every expression, such that Clang.jl prints them
+on top of the expression. The expressions themselves get sent to `rewriter(::Expr)`` for
+further treatment.
+"""
+function rewriter(xs::Vector)
+    rewritten = Any[]
+    for x in xs
+        # Clang.jl inserts strings like "# Skipping MacroDefinition: X"
+        # keep these to get a sense of what we are missing
+        if x isa String
+            push!(rewritten, x)
+            continue
+        end
+        @assert x isa Expr
+
+        x2, argpos = rewriter(x)
+        name = cname(x)
+        node = findnode(name, doc)
+        docstr = node === nothing ? "" : build_docstring(node, argpos)
+        isempty(docstr) || push!(rewritten, addquotes(docstr))
+        push!(rewritten, x2)
+    end
+    rewritten
+end
+
+"Make the arg at position i a keyword and move it to the back in the argpos permutation"
+function keywordify!(fargs2, argpos, i)
+    if i === nothing
+        return nothing
+    else
+        arg = fargs2[i]
+        fargs2[i] = :($arg = C_NULL)
+        # in optpos is does not have to be at i anymore if it already was moved
+        argoptpos = findfirst(==(i), argpos)
+        splice!(argpos, argoptpos)
+        push!(argpos, i)  # add it to the end
+    end
+end
+
+"Rewrite expressions in the ways listed at the top of this file."
```

Comment:
nit: Rewrite expressions using the transformations listed at the top of this file.